File: frontend/src/lib/pathResolver.ts
/**
 * Path Resolution for XSG Pattern Files
 *
 * Supports 4 path formats:
 * 1. @/ - Project-relative paths (e.g., "@/images/test.png")
 * 2. Relative paths (e.g., "../images/test.png")
 * 3. Absolute paths (e.g., "/home/user/images/test.png")
 * 4. URLs (e.g., "https://example.com/test.png")
 *
 * Current directory for relative paths: YAML file's directory
 * Project root detection: package.json, pyproject.toml, .git/, or cwd
 */

/**
 * Path resolution options
 */
export interface PathResolverOptions {
  /** Current YAML file path (used for relative path resolution) */
  currentFilePath?: string;
  /** Project root directory (optional, auto-detected if not provided) */
  projectRoot?: string;
  /** Base URL for development server (default: current origin) */
  baseUrl?: string;
}

/**
 * Path resolver class
 */
export class PathResolver {
  private currentFileDir: string | null = null;
  private projectRoot: string | null = null;
  private baseUrl: string;

  constructor(options: PathResolverOptions = {}) {
    // Extract current file directory from file path
    if (options.currentFilePath) {
      this.currentFileDir = this.getDirectory(options.currentFilePath);
    }

    // Set project root (or detect it)
    this.projectRoot = options.projectRoot || this.detectProjectRoot();

    // Set base URL
    this.baseUrl = options.baseUrl || window.location.origin;
  }

  /**
   * Resolve a path to an absolute URL
   *
   * @param path - Path to resolve (can be @/, relative, absolute, or URL)
   * @returns Resolved absolute URL
   */
  resolve(path: string): string {
    // 1. URL: Return as-is
    if (this.isUrl(path)) {
      return path;
    }

    // 2. Project-relative (@/)
    if (path.startsWith("@/")) {
      return this.resolveProjectRelative(path);
    }

    // 3. Absolute path
    if (this.isAbsolutePath(path)) {
      return this.resolveAbsolute(path);
    }

    // 4. Relative path
    return this.resolveRelative(path);
  }

  /**
   * Check if path is a URL
   */
  private isUrl(path: string): boolean {
    return /^https?:\/\//i.test(path);
  }

  /**
   * Check if path is absolute
   * Windows: C:\..., D:\..., \\server\...
   * Unix: /...
   */
  private isAbsolutePath(path: string): boolean {
    // Windows absolute paths
    if (/^[a-zA-Z]:[\\/]/.test(path)) return true;
    if (/^\\\\/.test(path)) return true; // UNC paths

    // Unix absolute paths
    if (path.startsWith("/")) return true;

    return false;
  }

  /**
   * Resolve project-relative path (@/...)
   */
  private resolveProjectRelative(path: string): string {
    // Remove @/ prefix
    const relativePath = path.slice(2);

    // In browser context, map to base URL
    // Backend will handle actual file system resolution
    if (this.projectRoot) {
      return `${this.baseUrl}/${relativePath}`;
    }

    // Fallback: treat as relative to base URL
    return `${this.baseUrl}/${relativePath}`;
  }

  /**
   * Resolve absolute path
   * In browser context, we need to convert to URL via backend API
   */
  private resolveAbsolute(path: string): string {
    // In browser, absolute file paths need to go through backend
    // For now, we'll use a file API endpoint
    const encodedPath = encodeURIComponent(path);
    return `${this.baseUrl}/api/file?path=${encodedPath}`;
  }

  /**
   * Resolve relative path
   * Relative to current YAML file's directory
   */
  private resolveRelative(path: string): string {
    if (!this.currentFileDir) {
      // No current file context, treat as relative to base URL
      return `${this.baseUrl}/${path}`;
    }

    // Normalize path separators
    const normalizedPath = path.replace(/\\/g, "/");

    // Combine current directory with relative path
    const combined = this.combinePaths(this.currentFileDir, normalizedPath);

    return `${this.baseUrl}/${combined}`;
  }

  /**
   * Get directory from file path
   */
  private getDirectory(filePath: string): string {
    // Handle URLs
    if (this.isUrl(filePath)) {
      const url = new URL(filePath);
      const pathParts = url.pathname.split("/");
      pathParts.pop(); // Remove filename
      return pathParts.join("/");
    }

    // Handle file paths
    const normalized = filePath.replace(/\\/g, "/");
    const lastSlash = normalized.lastIndexOf("/");

    if (lastSlash === -1) {
      return ""; // No directory, just filename
    }

    return normalized.slice(0, lastSlash);
  }

  /**
   * Combine two paths, handling .. and . segments
   */
  private combinePaths(basePath: string, relativePath: string): string {
    // Split paths into segments
    const baseSegments = basePath.split("/").filter((s) => s && s !== ".");
    const relativeSegments = relativePath
      .split("/")
      .filter((s) => s && s !== ".");

    // Process relative segments
    for (const segment of relativeSegments) {
      if (segment === "..") {
        // Go up one level
        if (baseSegments.length > 0) {
          baseSegments.pop();
        }
      } else {
        // Add segment
        baseSegments.push(segment);
      }
    }

    return baseSegments.join("/");
  }

  /**
   * Detect project root directory
   * Looks for: package.json, pyproject.toml, .git/
   * In browser context, this returns null (backend handles it)
   */
  private detectProjectRoot(): string | null {
    // In browser, we can't access file system directly
    // Backend will handle project root detection
    return null;
  }

  /**
   * Update current file path (for when loading a new YAML file)
   */
  setCurrentFilePath(filePath: string): void {
    this.currentFileDir = this.getDirectory(filePath);
  }

  /**
   * Update project root
   */
  setProjectRoot(projectRoot: string): void {
    this.projectRoot = projectRoot;
  }
}

/**
 * Default path resolver instance
 */
let defaultResolver: PathResolver | null = null;

/**
 * Get or create default path resolver
 */
export function getPathResolver(options?: PathResolverOptions): PathResolver {
  if (!defaultResolver) {
    defaultResolver = new PathResolver(options);
  } else if (options) {
    // Update options if provided
    if (options.currentFilePath) {
      defaultResolver.setCurrentFilePath(options.currentFilePath);
    }
    if (options.projectRoot) {
      defaultResolver.setProjectRoot(options.projectRoot);
    }
  }
  return defaultResolver;
}

/**
 * Convenience function to resolve a path using the default resolver
 */
export function resolvePath(
  path: string,
  options?: PathResolverOptions
): string {
  const resolver = getPathResolver(options);
  return resolver.resolve(path);
}

/**
 * Parse calc() expressions in coordinate strings
 * Examples:
 *   "50%" -> { type: 'percentage', value: 50 }
 *   "calc(50% + 10px)" -> { type: 'calc', expr: '50% + 10px' }
 *   100 -> { type: 'absolute', value: 100 }
 */
export type CoordinateValue =
  | { type: "absolute"; value: number }
  | { type: "percentage"; value: number }
  | { type: "calc"; expr: string };

export function parseCoordinate(coord: number | string): CoordinateValue {
  // Number: absolute pixels
  if (typeof coord === "number") {
    return { type: "absolute", value: coord };
  }

  // String: percentage or calc()
  const coordStr = coord.trim();

  // Percentage
  if (coordStr.endsWith("%")) {
    const value = parseFloat(coordStr);
    if (!isNaN(value)) {
      return { type: "percentage", value };
    }
  }

  // calc() expression
  const calcMatch = coordStr.match(/^calc\((.+)\)$/);
  if (calcMatch) {
    return { type: "calc", expr: calcMatch[1] };
  }

  // Fallback: treat as absolute
  const value = parseFloat(coordStr);
  return { type: "absolute", value: isNaN(value) ? 0 : value };
}

/**
 * Evaluate coordinate value to pixels
 *
 * @param coord - Coordinate value
 * @param containerSize - Container size (for percentage calculations)
 * @returns Pixel value
 */
export function evaluateCoordinate(
  coord: number | string,
  containerSize: number
): number {
  const parsed = parseCoordinate(coord);

  switch (parsed.type) {
    case "absolute":
      return parsed.value;

    case "percentage":
      return (parsed.value / 100) * containerSize;

    case "calc":
      // Simple calc() evaluation
      // Supports: 50% + 10px, 50% - 10px, etc.
      return evaluateCalcExpression(parsed.expr, containerSize);

    default:
      return 0;
  }
}

/**
 * Evaluate calc() expression
 * Supports basic arithmetic: +, -, *, /
 * Supports: px, %, numbers
 */
function evaluateCalcExpression(expr: string, containerSize: number): number {
  // Replace percentages with pixel values
  const normalized = expr.replace(/(\d+(?:\.\d+)?)%/g, (_match, num) => {
    const percentage = parseFloat(num);
    const pixels = (percentage / 100) * containerSize;
    return pixels.toString();
  });

  // Remove 'px' units
  const withoutUnits = normalized.replace(/px/g, "");

  // Evaluate the now purely-arithmetic expression with a sandboxed parser.
  // This intentionally avoids eval()/Function so that a hostile pattern file
  // cannot inject code through coordinate expressions.
  return evaluateArithmetic(withoutUnits);
}

/**
 * Safely evaluate a pure arithmetic expression to a number.
 *
 * This replaces the previous eval()-based evaluation to remove the code
 * injection risk: only numbers, the binary operators `+ - * /`, unary `+`/`-`,
 * parentheses, and whitespace are accepted. Any identifier, function call,
 * or stray symbol causes the whole expression to be rejected.
 *
 * Numeric semantics match JavaScript arithmetic: floating-point division,
 * `* /` bind tighter than `+ -`, operators are left-associative, and unary
 * minus is supported. On any tokenization/parse failure (unknown character,
 * empty or incomplete expression, etc.) this returns 0, matching the old
 * `catch { return 0 }` behavior.
 *
 * @param expr - Normalized arithmetic expression (no units, no percentages)
 * @returns Evaluated value, or 0 if the expression is invalid
 */
function evaluateArithmetic(expr: string): number {
  type Token =
    | { kind: "number"; value: number }
    | { kind: "op"; value: "+" | "-" | "*" | "/" }
    | { kind: "lparen" }
    | { kind: "rparen" };

  // --- Tokenizer ---------------------------------------------------------
  const tokenize = (input: string): Token[] | null => {
    const tokens: Token[] = [];
    let i = 0;

    while (i < input.length) {
      const ch = input[i];

      // Whitespace is ignored.
      if (ch === " " || ch === "\t" || ch === "\n" || ch === "\r") {
        i++;
        continue;
      }

      // Numbers: integer or decimal (e.g. "640", "0.5", ".5", "12.").
      if ((ch >= "0" && ch <= "9") || ch === ".") {
        let j = i;
        let seenDot = false;
        while (j < input.length) {
          const c = input[j];
          if (c >= "0" && c <= "9") {
            j++;
          } else if (c === "." && !seenDot) {
            seenDot = true;
            j++;
          } else {
            break;
          }
        }
        const slice = input.slice(i, j);
        const value = Number(slice);
        // Guard against malformed numbers like a bare ".".
        if (!Number.isFinite(value)) {
          return null;
        }
        tokens.push({ kind: "number", value });
        i = j;
        continue;
      }

      if (ch === "+" || ch === "-" || ch === "*" || ch === "/") {
        tokens.push({ kind: "op", value: ch });
        i++;
        continue;
      }

      if (ch === "(") {
        tokens.push({ kind: "lparen" });
        i++;
        continue;
      }

      if (ch === ")") {
        tokens.push({ kind: "rparen" });
        i++;
        continue;
      }

      // Anything else (letters, ';', ',', etc.) is rejected outright.
      return null;
    }

    return tokens;
  };

  // --- Recursive-descent parser ------------------------------------------
  // Grammar (left-associative, standard precedence):
  //   expression := term (('+' | '-') term)*
  //   term       := factor (('*' | '/') factor)*
  //   factor     := ('+' | '-') factor | '(' expression ')' | number
  const tokens = tokenize(expr);
  if (tokens === null) {
    return 0;
  }

  let pos = 0;
  const peek = (): Token | undefined => tokens[pos];

  // ParseError sentinel: thrown on any structural problem, caught below.
  class ParseError extends Error {}

  const parseFactor = (): number => {
    const token = peek();
    if (token === undefined) {
      throw new ParseError("unexpected end of expression");
    }

    // Unary +/-
    if (token.kind === "op" && (token.value === "+" || token.value === "-")) {
      pos++;
      const operand = parseFactor();
      return token.value === "-" ? -operand : operand;
    }

    if (token.kind === "lparen") {
      pos++;
      const value = parseExpression();
      const closing = peek();
      if (closing === undefined || closing.kind !== "rparen") {
        throw new ParseError("missing closing parenthesis");
      }
      pos++;
      return value;
    }

    if (token.kind === "number") {
      pos++;
      return token.value;
    }

    throw new ParseError("expected number or '('");
  };

  const parseTerm = (): number => {
    let value = parseFactor();
    for (;;) {
      const token = peek();
      if (token === undefined || token.kind !== "op") {
        break;
      }
      if (token.value !== "*" && token.value !== "/") {
        break;
      }
      pos++;
      const right = parseFactor();
      value = token.value === "*" ? value * right : value / right;
    }
    return value;
  };

  const parseExpression = (): number => {
    let value = parseTerm();
    for (;;) {
      const token = peek();
      if (token === undefined || token.kind !== "op") {
        break;
      }
      if (token.value !== "+" && token.value !== "-") {
        break;
      }
      pos++;
      const right = parseTerm();
      value = token.value === "+" ? value + right : value - right;
    }
    return value;
  };

  try {
    const result = parseExpression();
    // Reject trailing garbage (e.g. "1 2" or "(1)2").
    if (pos !== tokens.length) {
      return 0;
    }
    return Number.isFinite(result) ? result : 0;
  } catch {
    return 0;
  }
}

File: backend-python/app/path_resolver.py
"""
Path Resolution for XSG Pattern Files

Supports 4 path formats:
1. @/ - Project-relative paths (e.g., "@/images/test.png")
2. Relative paths (e.g., "../images/test.png")
3. Absolute paths (e.g., "/home/user/images/test.png")
4. URLs (e.g., "https://example.com/test.png")

Current directory for relative paths: YAML file's directory
Project root detection: package.json, pyproject.toml, .git/, or cwd
"""

import os
import re
from pathlib import Path
from typing import Optional, Union
from urllib.parse import urlparse


class PathResolver:
    """Path resolver for XSG patterns"""

    def __init__(
        self,
        current_file_path: Optional[str] = None,
        project_root: Optional[str] = None,
    ):
        """
        Initialize path resolver

        Args:
            current_file_path: Current YAML file path (for relative resolution)
            project_root: Project root directory (auto-detected if not provided)
        """
        self.current_file_dir: Optional[Path] = None
        if current_file_path:
            self.set_current_file_path(current_file_path)

        self.project_root: Optional[Path] = None
        if project_root:
            self.project_root = Path(project_root).resolve()
        else:
            self.project_root = self._detect_project_root()

    def resolve(self, path: str) -> str:
        """
        Resolve a path to an absolute path or URL

        Args:
            path: Path to resolve (can be @/, relative, absolute, or URL)

        Returns:
            Resolved absolute path or URL
        """
        # 1. URL: Return as-is
        if self._is_url(path):
            return path

        # 2. Project-relative (@/)
        if path.startswith("@/"):
            return self._resolve_project_relative(path)

        # 3. Absolute path
        if self._is_absolute_path(path):
            return str(Path(path).resolve())

        # 4. Relative path
        return self._resolve_relative(path)

    def _is_url(self, path: str) -> bool:
        """Check if path is a URL"""
        parsed = urlparse(path)
        return parsed.scheme in ("http", "https")

    def _is_absolute_path(self, path: str) -> bool:
        """
        Check if path is absolute

        Windows: C:\\..., D:\\..., \\\\server\\...
        Unix: /...
        """
        # Use pathlib for cross-platform absolute path detection
        return Path(path).is_absolute()

    def _resolve_project_relative(self, path: str) -> str:
        """
        Resolve project-relative path (@/...)

        Args:
            path: Path starting with @/

        Returns:
            Absolute file path
        """
        # Remove @/ prefix
        relative_path = path[2:]

        if self.project_root:
            resolved = self.project_root / relative_path
            return str(resolved.resolve())

        # Fallback: treat as relative to cwd
        resolved = Path.cwd() / relative_path
        return str(resolved.resolve())

    def _resolve_relative(self, path: str) -> str:
        """
        Resolve relative path
        Relative to current YAML file's directory

        Args:
            path: Relative path

        Returns:
            Absolute file path
        """
        if not self.current_file_dir:
            # No current file context, treat as relative to cwd
            resolved = Path.cwd() / path
            return str(resolved.resolve())

        # Combine current directory with relative path
        resolved = self.current_file_dir / path
        return str(resolved.resolve())

    def _detect_project_root(self) -> Optional[Path]:
        """
        Detect project root directory

        Looks for (in order):
        1. package.json (frontend)
        2. pyproject.toml (backend)
        3. .git/ (git repository)
        4. cwd (fallback)

        Returns:
            Project root path
        """
        current = Path.cwd()

        # Check current and parent directories
        for _ in range(10):  # Max 10 levels up
            # Check for markers
            if (current / "package.json").exists():
                return current
            if (current / "pyproject.toml").exists():
                return current
            if (current / ".git").is_dir():
                return current

            # Move up one level
            parent = current.parent
            if parent == current:
                # Reached filesystem root
                break
            current = parent

        # Fallback: cwd
        return Path.cwd()

    def set_current_file_path(self, file_path: str) -> None:
        """
        Update current file path

        Args:
            file_path: Path to current YAML file
        """
        if self._is_url(file_path):
            # For URLs, can't determine directory in traditional sense
            # Store as-is for potential URL-based relative resolution
            self.current_file_dir = None
        else:
            path = Path(file_path).resolve()
            self.current_file_dir = path.parent

    def set_project_root(self, project_root: str) -> None:
        """
        Update project root

        Args:
            project_root: Project root directory path
        """
        self.project_root = Path(project_root).resolve()


# Global default resolver
_default_resolver: Optional[PathResolver] = None


def get_path_resolver(
    current_file_path: Optional[str] = None,
    project_root: Optional[str] = None,
) -> PathResolver:
    """
    Get or create default path resolver

    Args:
        current_file_path: Current YAML file path
        project_root: Project root directory

    Returns:
        PathResolver instance
    """
    global _default_resolver

    if _default_resolver is None:
        _default_resolver = PathResolver(
            current_file_path=current_file_path,
            project_root=project_root,
        )
    else:
        # Update options if provided
        if current_file_path:
            _default_resolver.set_current_file_path(current_file_path)
        if project_root:
            _default_resolver.set_project_root(project_root)

    return _default_resolver


def resolve_path(
    path: str,
    current_file_path: Optional[str] = None,
    project_root: Optional[str] = None,
) -> str:
    """
    Convenience function to resolve a path using the default resolver

    Args:
        path: Path to resolve
        current_file_path: Current YAML file path
        project_root: Project root directory

    Returns:
        Resolved absolute path or URL
    """
    resolver = get_path_resolver(
        current_file_path=current_file_path,
        project_root=project_root,
    )
    return resolver.resolve(path)


def parse_coordinate(coord: Union[int, float, str]) -> dict:
    """
    Parse coordinate value

    Args:
        coord: Coordinate (number or string with %, calc())

    Returns:
        Dictionary with type and value/expr
    """
    # Number: absolute pixels
    if isinstance(coord, (int, float)):
        return {"type": "absolute", "value": coord}

    # String: percentage or calc()
    coord_str = str(coord).strip()

    # Percentage
    if coord_str.endswith("%"):
        try:
            value = float(coord_str[:-1])
            return {"type": "percentage", "value": value}
        except ValueError:
            pass

    # calc() expression
    calc_match = re.match(r"^calc\((.+)\)$", coord_str)
    if calc_match:
        return {"type": "calc", "expr": calc_match.group(1)}

    # Fallback: treat as absolute
    try:
        value = float(coord_str)
        return {"type": "absolute", "value": value}
    except ValueError:
        return {"type": "absolute", "value": 0}


def evaluate_coordinate(coord: Union[int, float, str], container_size: float) -> float:
    """
    Evaluate coordinate value to pixels

    Args:
        coord: Coordinate value
        container_size: Container size (for percentage calculations)

    Returns:
        Pixel value
    """
    parsed = parse_coordinate(coord)

    if parsed["type"] == "absolute":
        return parsed["value"]

    elif parsed["type"] == "percentage":
        return (parsed["value"] / 100) * container_size

    elif parsed["type"] == "calc":
        # Simple calc() evaluation
        return _evaluate_calc_expression(parsed["expr"], container_size)

    return 0


def _evaluate_calc_expression(expr: str, container_size: float) -> float:
    """
    Evaluate calc() expression

    Supports basic arithmetic: +, -, *, /
    Supports: px, %, numbers

    Args:
        expr: Calc expression
        container_size: Container size for percentage calculations

    Returns:
        Evaluated pixel value
    """
    # Replace percentages with pixel values
    def replace_percentage(match):
        percentage = float(match.group(1))
        pixels = (percentage / 100) * container_size
        return str(pixels)

    normalized = re.sub(r"(\d+(?:\.\d+)?)%", replace_percentage, expr)

    # Remove 'px' units
    without_units = normalized.replace("px", "")

    # Evaluate expression (simple evaluation)
    # WARNING: Using eval() - in production, use a proper expression parser
    try:
        # Restrict to basic arithmetic for safety
        if not re.match(r'^[\d\s+\-*/().]+$', without_units):
            return 0
        return float(eval(without_units))
    except Exception:
        return 0
